crates/ruff/src/rules/ruff/rules/unnecessary_list_allocation_for_first_element.rs

```diff
@@ -0,0 +1,198 @@
+                return None;
+            };
+
+            Some(arg_name.as_str())
+        }
+        Expr::ListComp(ast::ExprListComp {
+            elt, generators, ..
+        }) => {
+            // If the `elt` field is anything other than a [`Expr::Name`], we can't be sure that it
+            // doesn't modify the elements of the underlying iterator - for example, `[i + 1 for i in x][0]`.
+            if !matches!(elt.as_ref(), Expr::Name(ast::ExprName { .. })) {
+                return None;
+            }
+
+            // If there's more than 1 generator, we can't safely say that it fits the diagnostic conditions -
+            // for example, `[(i, j) for i in x for j in y][0]`.
+            if generators.len() != 1 {
+                return None;
+            }
+
+            let generator = &generators[0];
+            // Ignore if there's an `if` statement in the comprehension, since it filters the list.
+            if !generator.ifs.is_empty() {
+                return None;
+            }
+
+            let Some(arg_name) = get_name_id(&generator.iter) else {
+                return None;
+            };
+            Some(arg_name)
+        }
+        _ => None,
+    }
+}
+
+/// Check that the slice [`Expr`] is functionally equivalent to slicing into the first element. The
+/// first `bool` checks that the element is in fact first, the second checks if it's a slice or an
+/// index.
+fn classify_subscript(expr: &Expr) -> (bool, bool) {
+    match expr {
+        Expr::Constant(ast::ExprConstant { .. }) => {
+            let effective_index = get_effective_index(expr);
+            (acts_as_zero(effective_index), false)
+        }
+        Expr::Slice(ast::ExprSlice {
+            step: step_value,
+            lower: lower_index,
+            upper: upper_index,
+            ..
+        }) => {
+            let lower = lower_index.as_ref().and_then(|l| get_effective_index(l));
+            let upper = upper_index.as_ref().and_then(|u| get_effective_index(u));
+            let step = step_value.as_ref().and_then(|s| get_effective_index(s));
+
+            let in_slice = upper.is_some() || step.is_some();
+            if acts_as_zero(lower) {
+                if upper.unwrap_or(i64::MAX) > step.unwrap_or(1i64) {
+                    return (false, in_slice);
+                }
+
+                return (true, in_slice);
+            }
+
+            (false, in_slice)
+        }
+        _ => (false, false),
+    }
+}
+
+fn get_effective_index(expr: &Expr) -> Option<i64> {
+    match expr {
+        Expr::Constant(ast::ExprConstant {
+            value: Constant::Int(value),
+            ..
+        }) => value.to_i64(),
+        _ => None,
+    }
+}
+
+fn get_name_id(expr: &Expr) -> Option<&str> {
+    match expr {
+        Expr::Name(ast::ExprName { id, .. }) => Some(id),
+        _ => None,
+    }
+}
+
+fn acts_as_zero(i: Option<i64>) -> bool {
+    i.is_none() || i == Some(0i64)
```

Comment:
nit: Do we need a separate function for this? Also, I don't think you need to add a type annotation here.

```suggestion
    matches!(i, Some(0))
```